File: prismpath-telemetry-rs/src/decode.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Crystal Warden Supply Chain Labs LLC
//! Human-readable inspect path — decoding a captured telemetry bitstream against its flow.

use crate::quantizer::{self, FieldPartition, OTHER};
use crate::wire;
use crate::zeckendorf;
use prismpath_rs::{Graph, V};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

pub fn v_to_json(v: &V) -> serde_json::Value {
    match v {
        V::Null => serde_json::Value::Null,
        V::Bool(b) => serde_json::Value::Bool(*b),
        V::Num(n) => {
            if let Some(num) = serde_json::Number::from_f64(*n) {
                serde_json::Value::Number(num)
            } else if n.fract() == 0.0 {
                serde_json::Value::Number((*n as i64).into())
            } else {
                serde_json::Value::Null
            }
        }
        V::Str(s) => serde_json::Value::String(s.clone()),
        V::List(l) => serde_json::Value::Array(l.iter().map(v_to_json).collect()),
        V::Obj(entries) => {
            let mut map = serde_json::Map::new();
            for (k, val) in entries {
                map.insert(k.clone(), v_to_json(val));
            }
            serde_json::Value::Object(map)
        }
        V::Ellipsis => serde_json::Value::String("...".to_string()),
    }
}

fn serialize_v_map<S>(
    map: &HashMap<String, V>,
    serializer: S,
) -> Result<S::Ok, S::Error>
where
    S: serde::Serializer,
{
    let json_map: HashMap<String, serde_json::Value> = map
        .iter()
        .map(|(k, v)| (k.clone(), v_to_json(v)))
        .collect();
    json_map.serialize(serializer)
}

fn deserialize_v_map<'de, D>(
    deserializer: D,
) -> Result<HashMap<String, V>, D::Error>
where
    D: serde::Deserializer<'de>,
{
    let json_map: HashMap<String, serde_json::Value> = HashMap::deserialize(deserializer)?;
    Ok(json_map
        .into_iter()
        .map(|(k, v)| (k, V::from_json(&v)))
        .collect())
}

pub fn encode_readings(
    parts: &HashMap<String, FieldPartition>,
    readings: &[HashMap<String, V>],
) -> Result<String, String> {
    let mut out = String::new();
    for r in readings {
        out.push_str(&wire::encode_reading(parts, r)?);
    }
    Ok(out)
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReadingRow {
    pub symbols: HashMap<String, usize>,
    #[serde(serialize_with = "serialize_v_map", deserialize_with = "deserialize_v_map")]
    pub reading: HashMap<String, V>,
    pub routes: HashMap<String, Option<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InspectReport {
    pub fields: Vec<String>,
    pub decision_nodes: Vec<String>,
    pub n_readings: usize,
    pub trailing_ints: usize,
    pub readings: Vec<ReadingRow>,
    pub route_distribution: HashMap<String, usize>,
}

pub fn inspect(graph: &Graph, bits: &str) -> InspectReport {
    let parts = quantizer::build_partitions(graph);
    let fields = wire::order(&parts);
    let nodes = wire::decision_nodes(graph);
    let nf = fields.len();
    let wire_ints = zeckendorf::decode_stream(bits);
    let n_complete = wire_ints.len().checked_div(nf).unwrap_or(0);

    let mut rows = Vec::new();
    let mut route_dist: HashMap<String, usize> = HashMap::new();

    for i in 0..n_complete {
        let syms: HashMap<String, usize> = fields
            .iter()
            .enumerate()
            .map(|(j, f)| (f.clone(), wire_ints[i * nf + j] - 1))
            .collect();
        let reading = quantizer::reconstruct(&parts, &syms);
        let routes: HashMap<String, Option<String>> = nodes
            .iter()
            .map(|n| (n.clone(), wire::route_node(graph, n, &reading)))
            .collect();

        let show_reading: HashMap<String, V> = reading
            .into_iter()
            .map(|(f, v)| {
                if v == V::Str(OTHER.to_string()) {
                    (f, V::Str("<other>".to_string()))
                } else {
                    (f, v)
                }
            })
            .collect();

        let mut route_pairs: Vec<String> = nodes
            .iter()
            .map(|n| {
                let target = routes.get(n).cloned().flatten().unwrap_or_else(|| "None".to_string());
                format!("{}={}", n, target)
            })
            .collect();
        route_pairs.sort();
        let dist_key = route_pairs.join(" ");
        *route_dist.entry(dist_key).or_default() += 1;

        rows.push(ReadingRow {
            symbols: syms,
            reading: show_reading,
            routes,
        });
    }

    let trailing_ints = wire_ints.len() - n_complete * nf;

    InspectReport {
        fields,
        decision_nodes: nodes,
        n_readings: rows.len(),
        trailing_ints,
        readings: rows,
        route_distribution: route_dist,
    }
}

pub fn render(rep: &InspectReport) -> String {
    let mut out = vec![
        format!("fields: {:?}   decision nodes: {:?}", rep.fields, rep.decision_nodes),
        format!(
            "decoded {} reading(s){}",
            rep.n_readings,
            if rep.trailing_ints > 0 {
                format!("  (+{} trailing int(s) = partial final frame)", rep.trailing_ints)
            } else {
                "".to_string()
            }
        ),
        "".to_string(),
    ];

    for (i, r) in rep.readings.iter().enumerate() {
        let mut route_parts = Vec::new();
        for n in &rep.decision_nodes {
            if let Some(target) = r.routes.get(n) {
                route_parts.push(format!(
                    "{}->{}",
                    n,
                    target.as_deref().unwrap_or("None")
                ));
            }
        }
        let route_str = route_parts.join(", ");
        out.push(format!("#{:4}  {:?}   =>  {}", i, r.reading, route_str));
    }

    out.push("".to_string());
    out.push("route distribution:".to_string());

    let mut dist_entries: Vec<(&String, &usize)> = rep.route_distribution.iter().collect();
    dist_entries.sort_by_key(|&(_, count)| std::cmp::Reverse(*count));

    for (k, c) in dist_entries {
        out.push(format!("  {:6}  {}", c, k));
    }

    out.join("\n")
}

File: prismpath/telemetry/decode.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Crystal Warden Supply Chain Labs LLC
"""Human-readable inspect path  -  the answer to "opaque wire, no tcpdump".

Point it at a captured telemetry bitstream + the flow `.md` and it decodes each reading back to its
symbols, the reconstructed representative values, and  -  the useful part  -  the **routing decision** the
policy makes on it. The `.md` IS the decoder: it defines both the field partition (how bits become
symbols) and the routing (how symbols become a decision), so no bespoke, drifting tooling is needed.

Kept as a standalone adapter tool (no core change); a `prismpath decode` CLI alias would be a one-line
wiring in `prismpath/cli.py` if desired.

  python prismpath/telemetry/decode.py --flow <flow.md> --bits <stream-of-0s-and-1s | ->  [--json]
"""
from __future__ import annotations

import argparse
import json
import sys
from collections import Counter
from pathlib import Path
from typing import Dict, List

try:
    import prismpath  # noqa: F401  installed or already on the path: leave sys.path alone
except ImportError:  # run as a loose script from a clone: make the repo root importable
    sys.path.insert(0, str(Path(__file__).resolve().parent.parent.parent))

from prismpath.telemetry import quantizer   # noqa: E402
from prismpath.telemetry import wire        # noqa: E402
from prismpath.telemetry import zeckendorf as zeck  # noqa: E402
from prismpath.kernel.parser import parse  # noqa: E402


def _show(value):
    return "<other>" if value == quantizer.OTHER_CELL else value


def encode_readings(parts, readings: List[dict]) -> str:
    """The multi-reading wire: each reading's fields in canonical order, concatenated (zero header)."""
    return "".join(wire.encode_reading(parts, reading) for reading in readings)


def inspect(graph, bits: str) -> Dict:
    """Decode a bitstream against a flow -> per-reading symbols, reconstructed values, and routes."""
    parts = quantizer.build_partitions(graph)
    fields = sorted(parts.keys())
    nodes = wire.decision_nodes(graph)
    nf = len(fields)
    wire_ints = zeck.decode_stream(bits)
    n_complete = len(wire_ints) // nf if nf else 0
    rows = []
    for reading_index in range(n_complete):
        syms = {fields[field_index]: wire_ints[reading_index * nf + field_index] - 1
                for field_index in range(nf)}
        reading = quantizer.reconstruct(parts, syms)
        routes = {node: wire.route_node(graph, node, reading) for node in nodes}
        rows.append({"symbols": syms,
                     "reading": {field: _show(value) for field, value in reading.items()},
                     "routes": routes})
    trailing = len(wire_ints) - n_complete * nf          # leftover ints = a partial final frame
    dist = Counter(tuple(sorted(row["routes"].items())) for row in rows)
    return {"fields": fields, "decision_nodes": nodes, "n_readings": len(rows),
            "trailing_ints": trailing, "readings": rows,
            "route_distribution": {" ".join(f"{node}={target}" for node, target in route_pairs): count
                                   for route_pairs, count in dist.items()}}


def _render(rep: Dict) -> str:
    out = [f"fields: {rep['fields']}   decision nodes: {rep['decision_nodes']}",
           f"decoded {rep['n_readings']} reading(s)"
           + (f"  (+{rep['trailing_ints']} trailing int(s) = partial final frame)"
              if rep["trailing_ints"] else ""), ""]
    for reading_index, reading in enumerate(rep["readings"]):
        route = ", ".join(f"{node}->{target}" for node, target in reading["routes"].items())
        out.append(f"#{reading_index:4d}  {reading['reading']}   =>  {route}")
    out += ["", "route distribution:"]
    for route_label, count in sorted(rep["route_distribution"].items(), key=lambda kv: -kv[1]):
        out.append(f"  {count:6d}  {route_label}")
    return "\n".join(out)


def main(argv=None) -> int:
    ap = argparse.ArgumentParser(description="Decode a PrismPath telemetry bitstream against its flow.")
    ap.add_argument("--flow", required=True, help="the flow .md (the decoder)")
    ap.add_argument("--bits", required=True, help="file of 0/1 chars, or - for stdin")
    ap.add_argument("--json", action="store_true")
    args = ap.parse_args(argv)
    bits = sys.stdin.read() if args.bits == "-" else Path(args.bits).read_text()
    bits = "".join(ch for ch in bits if ch in "01")
    graph = parse(Path(args.flow).read_text())
    rep = inspect(graph, bits)
    print(json.dumps(rep, indent=1) if args.json else _render(rep))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
